contracts/cw20-staked-balance-voting/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+                    let msg = WasmMsg::Instantiate {
+                        code_id: staking_code_id,
+                        funds: vec![],
+                        admin: Some(info.sender.to_string()),
+                        label: env.contract.address.to_string(),
+                        msg: to_binary(&stake_cw20::msg::InstantiateMsg {
+                            admin: Some(info.sender.to_string()),
+                            unstaking_duration,
+                            token_address: address.to_string(),
+                        })?,
+                    };
+                    let msg = SubMsg::reply_on_success(msg, INSTANTIATE_STAKING_REPLY_ID);
+                    Ok(Response::default()
+                        .add_attribute("action", "instantiate")
+                        .add_attribute("token", "existing_token")
+                        .add_attribute("token_address", address)
+                        .add_submessage(msg))
+                }
+            }
+        }
+        TokenInfo::New {
+            code_id,
+            label,
+            name,
+            symbol,
+            decimals,
+            initial_balances,
+            marketing,
+            staking_code_id,
+            unstaking_duration,
+        } => {
+            let initial_supply = initial_balances
+                .iter()
+                .fold(Uint128::zero(), |p, n| p + n.amount);
+            if initial_supply.is_zero() {
+                return Err(ContractError::InitialBalancesError {});
+            }
+
+            STAKING_CONTRACT_CODE_ID.save(deps.storage, &staking_code_id)?;
+            STAKING_CONTRACT_UNSTAKING_DURATION.save(deps.storage, &unstaking_duration)?;
+
+            let msg = WasmMsg::Instantiate {
+                admin: Some(info.sender.to_string()),
+                code_id,
+                msg: to_binary(&cw20_base::msg::InstantiateMsg {
+                    name,
+                    symbol,
+                    decimals,
+                    initial_balances,
+                    mint: Some(cw20::MinterResponse {
+                        minter: info.sender.to_string(),
+                        cap: None,
+                    }),
+                    marketing,
+                })?,
+                funds: vec![],
+                label,
+            };
+            let msg = SubMsg::reply_on_success(msg, INSTANTIATE_TOKEN_REPLY_ID);
+
+            Ok(Response::default()
+                .add_attribute("action", "instantiate")
+                .add_attribute("token", "new_token")
+                .add_submessage(msg))
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    _deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {}
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
```

Comment:
What do you think about adding a query here to get the DAO that this contract is associated with? I don't see it being used terribly often but seems like it could be useful + we already have that info stored in state.